### PR TITLE
QA-1974: Run Rawls tests by tag. [Good for review]

### PR DESCRIPTION
**Pull request by ichengchang**
**Description**

JIRA ticket [QA-1974](https://broadworkbench.atlassian.net/browse/QA-1974). See also parent ticket [QA-1970](https://broadworkbench.atlassian.net/browse/QA-1970).

Group Rawls tests so they can be run by following tags (see [Rawls Swat Test Ownership](https://docs.google.com/spreadsheets/d/1YxghKs-16IRxJ6R6N9HoAKGc3I6ly2MTaCuMuAay1nE/edit#gid=0))

- AuthDomainsTest
- BillingsTest
- DataRepoSnapshotsTest
- MethodsTest
- RawlsTest
- WorkspacesTest

There is an associated PR in `dsp-jenkins` [#616](https://github.com/broadinstitute/dsp-jenkins/pull/616) that goes with this ticket.

The following new Jenkins job demonstrates how the router works for PRs.

- [https://fc-jenkins.dsp-techops.broadinstitute.org/job/swatomation-pipeline-rawls/](https://fc-jenkins.dsp-techops.broadinstitute.org/job/swatomation-pipeline-rawls/)
- [https://fc-jenkins.dsp-techops.broadinstitute.org/job/rawls-fiab-test-router/](https://fc-jenkins.dsp-techops.broadinstitute.org/job/rawls-fiab-test-router/)

Slack notifications will be handled in another ticket.

- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check that the **Product Owner** has signed off on any user-facing changes
- [ ] **Submitter**: Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [ ] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Database checks:
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] Tell your tech lead (TL) that the PR exists if they want to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

**Review by ungwudik — COMMENTED** [review]
The setup here looks good to me. But we need to change/update the tag names. I think the current naming scheme ties the tests to specific teams. That will create an additional overhead if team ownership for these tests change, which means we'll then have to update the tests to reflect that. Instead, can you name them based on the functionality? For example, instead of tagging `MethodConfigApiSpec` as `@BatchWorkflowsTest`, we can tag it as `@MethodsTest`. Then when we kick off the router job for BatchWorkflows team, we'll send the `@MethodsTest` tag (along with all other tags that BatchWorkflows team owns). If ownership for `MethodConfigApiSpec` changes, we'll just update the router job to start sending the failures to whoever the new team is.

P.S. We still have to make _some_ updates (in jenkins, but not both in jenkins and in the rawls tests themselves) if team ownership changes. The idea is to limit those changes as much as we can.

**Review by ungwudik — COMMENTED** [review]
I like the term `router job`, by the way :)

**Comment by ichengchang** [comment]
> 

Sounds good! I would have done the grouping based on the functionality in the first place. I hesitated only because it looks like these are still **Open Questions** in [QA-1970](https://broadworkbench.atlassian.net/browse/QA-1970). I think we should change the **Open Questions** section to **Action Items**.

**Comment by ichengchang** [comment]
> The setup here looks good to me. But we need to change/update the tag names. I think the current naming scheme ties the tests to specific teams. That will create an additional overhead if team ownership for these tests change, which means we'll then have to update the tests to reflect that. Instead, can you name them based on the functionality? For example, instead of tagging `MethodConfigApiSpec` as `@BatchWorkflowsTest`, we can tag it as `@MethodsTest`. Then when we kick off the router job for BatchWorkflows team, we'll send the `@MethodsTest` tag (along with all other tags that BatchWorkflows team owns). If ownership for `MethodConfigApiSpec` changes, we'll just update the router job to start sending the failures to whoever the new team is.
> 
> P.S. We still have to make _some_ updates (in jenkins, but not both in jenkins and in the rawls tests themselves) if team ownership changes. The idea is to limit those changes as much as we can.

The refactoring of tags into names closer to API functionality has been completed.

**Comment by ichengchang** [comment]
The `fiab-start` job failed for PRs multiple times but the hourly tests look good. I'll leave it just as that.

**Review by ungwudik — APPROVED** [review]
LGTM

- Nit - Can you update the description of this PR to reflect the new tag names?
- Also, can you add an engineer from each of the Rawls teams, and get their approval as well?

**Comment by ungwudik** [comment]
> The `fiab-start` job failed for PRs multiple times but the hourly tests look good. I'll leave it just as that.

I say let's still get a passing test run before merging. Although this PR is not particularly an intrusive change, it does touch all Rawls swatomation tests and I'd rather err on the side of caution.

**Comment by ichengchang** [comment]
> LGTM
> 
> * Nit - Can you update the description of this PR to reflect the new tag names?
> * Also, can you add an engineer from each of the Rawls teams, and get their approval as well?

Updated PR description of this PR.

**Comment by ichengchang** [comment]
> 

This is the latest PR run ([https://fc-jenkins.dsp-techops.broadinstitute.org/job/swatomation-pipeline/68904/](https://fc-jenkins.dsp-techops.broadinstitute.org/job/swatomation-pipeline/68904/)) shows all tests being successful through the new rawls router. 

The rawls test without the router (running all rawls tests) has 1 failed test case out of 79 total.

We could keep trying until the original rawls pass or say it's good enough as long as all routed tests passed. Up to you.

File: automation/src/test/scala/org/broadinstitute/dsde/test/api/tagannotation/AuthDomainsTest.java
Comment (ehigham) [inline automation/src/test/scala/org/broadinstitute/dsde/test/api/tagannotation/AuthDomainsTest.java]: Is it possible to combine these into one interface? Might help organisation and future discovery of these annotations?

Comment (ichengchang) [reply automation/src/test/scala/org/broadinstitute/dsde/test/api/tagannotation/AuthDomainsTest.java]: Great idea! I've refactored the annotations and group them into a higher-level `RawlsTestGroupAnnotation`.

Comment (ehigham) [reply automation/src/test/scala/org/broadinstitute/dsde/test/api/tagannotation/AuthDomainsTest.java]: Oh  I see how my comment was ambiguous. I meant combine them into one file rather than many. This is a nice improvement though!
```java
public interface RawlsTestGroupAnnotations
{
  @RawlsTestGroupAnnotation
  public @interface AuthDomainTest {}

  @RawlsTestGroupAnnotation
  public @interface BillingTest {}   

    // etc
}
```

**Comment by ungwudik** [comment]
What does `AuthDomainsTest1`, `BillingsTest1`, etc mean? I thought the tags (for example `AuthDomainsTest`, without a trailing `1`) are already represented in `RawlsTestGroupAnnotations.java`?

**Comment by ichengchang** [comment]
> What does `AuthDomainsTest1`, `BillingsTest1`, etc mean? I thought the tags (for example `AuthDomainsTest`, without a trailing `1`) are already represented in `RawlsTestGroupAnnotations.java`?

I renamed the original annotations in case I need to rollback. I am still testing the new `RawlsTestGroupAnnotations.java` with all annotations rolled into a single file, this hasn't been successful yet.

File: automation/src/test/scala/org/broadinstitute/dsde/test/api/tagannotation/rawls/RawlsTestTaggedAs.scala
Comment (aherbst-broad) [inline automation/src/test/scala/org/broadinstitute/dsde/test/api/tagannotation/rawls/RawlsTestTaggedAs.scala]: It'd be good to get rid of the `Obj1` suffix once testing is complete (if I'm understanding the upstream PR conversation correctly). Do you plan on doing that prior to merge?

**Review by aherbst-broad — COMMENTED** [review]
Had a question about the `Obj1` suffixes.

**Review by ehigham — APPROVED** [review]
Looks great!

File: automation/src/test/scala/org/broadinstitute/dsde/test/api/tagannotation/AuthDomainsTest.java
Comment (ichengchang) [reply automation/src/test/scala/org/broadinstitute/dsde/test/api/tagannotation/AuthDomainsTest.java]: Some updates here.

The `@interface` implicitly means `interface extends annotation.Annotation`. We can't extend `@interface` like a regular class or treat them like a regular java type. 

We can, however, place them in a non-class all in one file. The only access modifier allowed is `default` so the `@interface` must reside in the same package as the tests. I think this should be fine.

I have refactored the code according to my explanation above.

File: automation/src/test/scala/org/broadinstitute/dsde/test/api/tagannotation/rawls/RawlsTestTaggedAs.scala
Comment (ichengchang) [reply automation/src/test/scala/org/broadinstitute/dsde/test/api/tagannotation/rawls/RawlsTestTaggedAs.scala]: Yes. I have cleaned up and refactored the annotation tags. I managed to place all annotation tags in one file as well.

**Pull request merged**
